File: python3/2904.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        l, r = 0, 0
        ans = ""
        cnt = 0
        n = len(s)
        while r < n:
            cnt += 1 if s[r] == '1' else 0
            while cnt > k:
                cnt -= 1 if s[l] == '1' else 0
                l += 1
            while l <= r and s[l] == '0':
                l += 1
            if cnt == k:
                n2 = r - l + 1
                if ans == "" or len(ans) > n2:
                    ans = s[l : r + 1]
                elif len(ans) == n2:
                    for i in range(n2):
                        if ans[i] > s[l + i]:
                            ans = s[l: r + 1]
                            break
                        if ans[i] < s[l + i]:
                            break
            r += 1
        return ans

File: cpp/2904.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        int l = 0, r = 0;
        int n = s.size();
        string ans;
        int cnt = 0;
        while (r < n) {
            cnt += s[r] == '1' ? 1 : 0;
            while (cnt > k) {
                cnt -= s[l] == '1' ? 1 : 0;
                ++l;
            }
            while (s[l] == '0') {
                ++l;
            }
            if (cnt == k) { // 美丽字符串
                int len = r - l + 1;
                if (ans.size() == 0 || ans.size() > len) {  // 更短的美丽字符串
                    ans = s.substr(l, len);
                } else if (ans.size() == len) { // 判断哪个美丽字符串字典序更小
                    for(int i = 0; i < ans.size(); ++i) {
                        if (ans[i] > s[l + i]) {
                            ans = s.substr(l, len);
                            break;
                        }
                        if (ans[i] < s[l + i]) {
                            break;
                        }
                    }
                }
            }
            ++r;
        }
        return ans;
    }
};